File: backend/modules/integration/fireflies.py
import time
import httpx


class FirefliesError(Exception):
    pass


class FirefliesClient:
    def __init__(self, api_key: str, timeout_seconds: int = 20) -> None:
        self.api_key = api_key
        self.endpoint = "https://api.fireflies.ai/graphql"
        self.timeout_seconds = timeout_seconds

    @staticmethod
    def parse_transcript_id(link: str) -> str | None:
        if not link:
            return None
        from urllib.parse import urlparse

        try:
            parsed = urlparse(link)
        except Exception:
            return None
        if not parsed.path:
            return None
        parts = [part for part in parsed.path.split("/") if part]
        if not parts:
            return None
        last = parts[-1]
        if "::" in last:
            return last.split("::")[-1]
        return last

    def _post(self, query: str, variables: dict | None = None) -> dict:
        headers = {"Authorization": f"Bearer {self.api_key}"}
        payload = {"query": query, "variables": variables or {}}
        with httpx.Client(timeout=self.timeout_seconds) as client:
            response = client.post(self.endpoint, json=payload, headers=headers)
        if response.status_code >= 400:
            raise FirefliesError(f"HTTP {response.status_code}: {response.text}")
        data = response.json()
        if data.get("errors"):
            raise FirefliesError(str(data["errors"]))
        return data.get("data", {})

    def validate_key(self) -> bool:
        query = "query { user { email } }"
        try:
            data = self._post(query)
        except FirefliesError:
            return False
        return bool(data.get("user", {}).get("email"))

    def get_transcript(self, transcript_id: str) -> dict:
        query = """
        query Transcript($id: String!) {
          transcript(id: $id) {
            id
            title
            date
            duration
            meeting_link
            sentences {
              index
              speaker_name
              speaker_id
              text
              start_time
              end_time
            }
          }
        }
        """
        data = self._post(query, {"id": transcript_id})
        transcript = data.get("transcript")
        if not transcript:
            raise FirefliesError("Transcript not found")
        return transcript

    def poll_transcript(
        self,
        transcript_id: str,
        attempts: int = 5,
        delay_seconds: int = 10,
    ) -> dict:
        last_error: Exception | None = None
        for _ in range(attempts):
            try:
                return self.get_transcript(transcript_id)
            except Exception as exc:
                last_error = exc
                time.sleep(delay_seconds)
        raise FirefliesError(f"Transcript polling failed: {last_error}")

File: knack_v2/tools/knack-v2/src/lib/fireflies.ts
const FIREFLIES_GRAPHQL_URL = "https://api.fireflies.ai/graphql";
const FIREFLIES_HOSTS = new Set(["app.fireflies.ai", "fireflies.ai"]);

type FirefliesTranscriptSentence = {
  index?: number | null;
  speaker_name?: string | null;
  speaker_id?: string | null;
  text?: string | null;
  start_time?: number | null;
  end_time?: number | null;
};

type FirefliesTranscriptResponse = {
  id?: string | null;
  title?: string | null;
  date?: string | null;
  duration?: number | null;
  meeting_link?: string | null;
  sentences?: FirefliesTranscriptSentence[] | null;
};

export function parseTranscriptId(link: string): string {
  let url: URL;

  try {
    url = new URL(link);
  } catch {
    throw new Error("Invalid transcript link: expected a valid URL");
  }

  if (!FIREFLIES_HOSTS.has(url.hostname)) {
    throw new Error("Invalid transcript link: expected a Fireflies URL");
  }

  const segments = url.pathname.split("/").filter(Boolean);
  const viewIndex = segments.findIndex((segment) => segment === "view");
  const transcriptId = viewIndex >= 0 ? segments[viewIndex + 1] : segments.at(-1);

  if (!transcriptId) {
    throw new Error("Invalid transcript link: could not parse transcript id");
  }

  return transcriptId;
}

export async function fetchTranscript(apiKey: string, transcriptId: string): Promise<FirefliesTranscriptResponse> {
  const query = `
    query Transcript($id: String!) {
      transcript(id: $id) {
        id
        title
        date
        duration
        meeting_link
        sentences {
          index
          speaker_name
          speaker_id
          text
          start_time
          end_time
        }
      }
    }
  `;

  const response = await fetch(FIREFLIES_GRAPHQL_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      query,
      variables: { id: transcriptId },
    }),
  });

  if (!response.ok) {
    const body = await response.text();
    throw new Error(`Fireflies request failed: HTTP ${response.status} ${body}`);
  }

  const payload = await response.json();

  if (payload.errors?.length) {
    throw new Error(`Fireflies API error: ${JSON.stringify(payload.errors)}`);
  }

  const transcript = payload.data?.transcript;

  if (!transcript) {
    throw new Error("Fireflies transcript not found");
  }

  return transcript;
}

export function renderTranscriptText(transcript: FirefliesTranscriptResponse): string {
  const sentences = transcript.sentences ?? [];

  return sentences
    .map((sentence) => {
      const speaker = sentence.speaker_name?.trim() || "Speaker";
      const text = sentence.text?.trim() || "";
      return text ? `${speaker}: ${text}` : "";
    })
    .filter(Boolean)
    .join("\n");
}

export function cleanTranscriptText(text: string): string {
  return text
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean)
    .join("\n");
}
